### ISSUE TITLE
[Feature Request] AssetManager multi core loading support

### ISSUE THREAD

**Issue by cubesdev** [libgdx/libgdx#5770]
#### Issue details
My game (for desktop platform) takes a long time to load all the assets. I have done some profiling and testing and noticed that the loading time could be reduced a lot if multiple threads were used for asynchronous loading in AssetManager instead of only one. 

I tried to increase the number of AsyncExecutor threads. But this does not work since the current implementation always waits for the current task to compete before scheduling a new one. I tried to implement a custom AssetManager but this is not very easy to do because the AssetManager itself is passed as a reference to every AssetLoader. Creating and using an interface for AssetManager in loader class would help libGDX users to implement such a custom AssetManager. Currently it is not possible because AssetManager is hardly extendable (mostly final fields, package private stuff, ...)

#### Version of LibGDX and/or relevant dependencies
1.9.10

#### Please select the affected platforms
- [x] Android
- [x] iOS (robovm)
- [x] iOS (MOE)
- [x] HTML/GWT
- [x] Windows
- [x] Linux
- [x] MacOS

**Comment by NathanSweet** [comment]
Using multiple threads would only be possible for items whose dependencies are already loaded, or that have no dependencies. It would be a nice addition, but difficult as AssetManager is quite complicated already. Maybe you would be willing to submit a PR for it?

**Comment by mgsx-dev** [comment]
@cubesdev did you tried to change this hard coded value here https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/assets/AssetManager.java#L122 ? it seems that non GPU ressources are loaded async. with a higher number of threads.

Maybe a constructor parameter for this hard coded value could be a good addition.

**Comment by cubesdev** [comment]
@mgsx-dev Yes i tried that. The hard coded value does not change anything because there is only one task scheduled at once (the next one only gets scheduled when the previous task completed). So a constructor parameter does not solve the issue.

@NathanSweet Exactly. The AssetManager class (along with AssetLoadingTask) implementations are very complicated (I assume because "modern" Java classes like CompletableFuture were avoided to stay compatible with old Android versions or GWT?). I could submit a PR but I think that a lot of refactoring needs to be done and maybe even some AssetLoader classes need to adopt. Maybe we could create a clean interface class for AssetManager which is used in the AssetLoader classes. Then it would be possible to create a separate performant and modern AssetManager implementation (maybe even as a separate extension?) without changing the current one too much).

**Comment by NathanSweet** [comment]
The AssetManager we have is not needlessly complex for the most part and is well tested. Ideally we would implement multiple threads without rewriting most of it.

**Comment by cubesdev** [comment]
So you think that using an interface for AssetManager is not worth it? Do you have a solution in mind to support multiple threads with the current implementation?

**Comment by NathanSweet** [comment]
We don't need multiple AssetManager implementations, so I don't think an interface is needed. If we build an AssetManager with multiple threads, there's no point in using the old one.

The current implementation can be modified to support multiple threads. Instead of one, submit up to X tasks where X is the max number of threads in the pool. Only submit tasks that have all their dependencies loaded. There are many details to consider to do that. It is not a small project.

I don't think building an AssetManager that supports multiple threads will be done easier by starting over from scratch. It's not clear how much newer JRE classes would help. It's not just threading that is tricky to get right, but also dependencies and ref counting.

**Comment by cubesdev** [comment]
> We don't need multiple AssetManager implementations, so I don't think an interface is needed. If we build an AssetManager with multiple threads, there's no point in using the old one.

Yes, that's true. But getting those modifications done without breaking anything might be tricky. An interface would at least help to implement an experimental version. 

> The current implementation can be modified to support multiple threads. Instead of one, submit up to X tasks where X is the max number of threads in the pool. Only submit tasks that have all their dependencies loaded. There are many details to consider to do that. It is not a small project.

I have already tried some small changes and immediately got 2x loading speed. My game is starting without any visible loading problems. But my asset dependency graph is (mostly) a tree and not an arbitrary acyclic graph. I am not sure what happens if there are shared dependencies between assets, etc.

**Comment by cubesdev** [comment]
I also propose to allow configuring the AsyncExecutor implementation to use ExecutorService implementations with a zero core pool size and a defined keepAliveTime. Otherwise the AssetManager's thread is never shutdown even after long idle time. That's not a big issue but every thread consumes some memory. On top of that many simple games will only load assets at the beginning and won't take up the loading thread ever again.

**Comment by anbu93** [comment]
I try realize it, not tested yet. I used it for my game.
https://github.com/anbu93/libgdx/commit/5307b52474ae3f2e8295aa0fe571a6edcd3a3f27
